File: rust/tcl-registry/src/commands/irules/botdefense__micro_service.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `BOTDEFENSE::micro_service` iRules command.
use crate::prelude::*;
pub const fn spec() -> CommandSpec {
    CommandSpec {
        name: "BOTDEFENSE::micro_service",
        dialects: Some(DialectSet::IRULES),
        arity: Arity::at_least(0),
        hover: Some(HoverSnippet {
            summary: "Returns the micro-service that matched the current request.",
            synopsis: &["BOTDEFENSE::micro_service (name | type)"],
            snippet: "Returns the micro-service that matched the current request.",
            source: "https://clouddocs.f5.com/api/irules/BOTDEFENSE__micro_service.html",
            examples: "when BOTDEFENSE_ACTION {\n    set ms [BOTDEFENSE::micro_service name]\n    if { $ms neq \"\"} {\n        log.local0. \"Request to micro_service $ms of type [BOTDEFENSE::micro_service type]\n    }\n}",
            return_value: "Returns the name or type of the micro-service found for the current request",
        }),
        event_requires: Some(EventRequires {
            client_side: false,
            server_side: false,
            transport: None,
            profiles: &["BOTDEFENSE"],
            also_in: &[],
            init_only: false,
            flow: false,
            capability: None,
        }),
        forms: &[FormSpec {
            kind: FormKind::Default,
            synopsis: "BOTDEFENSE::micro_service (name | type)",
            dialects: None,
        }],
        side_effects: &[SideEffect {
            target: SideEffectTarget::AsmState,
            reads: true,
            writes: false,
            connection_side: ConnectionSide::Client,
            dialects: None,
        }],
        ..CommandSpec::DEFAULT
    }
}

File: core/commands/registry/irules/botdefense__micro_service.py
# Enriched from F5 iRules reference documentation.
"""BOTDEFENSE::micro_service -- Returns the micro-service that matched the current request."""

from __future__ import annotations

from ....compiler.side_effects import ConnectionSide, SideEffect, SideEffectTarget
from .._base import CommandDef, make_av
from ..models import CommandSpec, FormKind, FormSpec, HoverSnippet, ValidationSpec
from ..namespace_models import EventRequires
from ..signatures import Arity
from ._base import _IRULES_ONLY, register

_SOURCE = "https://clouddocs.f5.com/api/irules/BOTDEFENSE__micro_service.html"


_av = make_av(_SOURCE)


@register
class BotdefenseMicroServiceCommand(CommandDef):
    name = "BOTDEFENSE::micro_service"

    @classmethod
    def spec(cls) -> CommandSpec:
        return CommandSpec(
            name="BOTDEFENSE::micro_service",
            dialects=_IRULES_ONLY,
            hover=HoverSnippet(
                summary="Returns the micro-service that matched the current request.",
                synopsis=("BOTDEFENSE::micro_service (name | type)",),
                snippet="Returns the micro-service that matched the current request.",
                source=_SOURCE,
                examples=(
                    "when BOTDEFENSE_ACTION {\n"
                    "    set ms [BOTDEFENSE::micro_service name]\n"
                    '    if { $ms neq ""} {\n'
                    '        log.local0. "Request to micro_service $ms of type [BOTDEFENSE::micro_service type]\n'
                    "    }\n"
                    "}"
                ),
                return_value="Returns the name or type of the micro-service found for the current request",
            ),
            forms=(
                FormSpec(
                    kind=FormKind.DEFAULT,
                    synopsis="BOTDEFENSE::micro_service (name | type)",
                    arg_values={
                        0: (
                            _av(
                                "name",
                                "BOTDEFENSE::micro_service name",
                                "BOTDEFENSE::micro_service (name | type)",
                            ),
                            _av(
                                "type",
                                "BOTDEFENSE::micro_service type",
                                "BOTDEFENSE::micro_service (name | type)",
                            ),
                        )
                    },
                ),
            ),
            validation=ValidationSpec(
                arity=Arity(),
            ),
            event_requires=EventRequires(profiles=frozenset({"BOTDEFENSE"})),
            side_effect_hints=(
                SideEffect(
                    target=SideEffectTarget.ASM_STATE,
                    reads=True,
                    connection_side=ConnectionSide.CLIENT,
                ),
            ),
        )
